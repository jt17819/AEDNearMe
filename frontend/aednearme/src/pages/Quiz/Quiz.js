import { Button, FormControl, InputLabel, MenuItem, Paper, Select } from '@material-ui/core';
import React, {useState} from 'react'
import axios from 'axios'

const Quiz = () => {
    let data 
    const baseUrl = 'http://localhost:8000/quiz/'
    const [category, setCategory] = useState('CPR')
    const [questions, setQuestions] = useState([])
    const [correctAnswer, setCorrectAnswer] = useState([])
    const [options, setOptions] = useState([])
    const [score, setScore] = useState([0,0,0,0,0,0])
    const [finalScore,setFinalScore] = useState(0)


    const handleCategory = (e) => setCategory(e.target.value)

    const handleScore = (e) => {
      for (let i=0; i<correctAnswer.length; i++)  {
        const newScore = score
        if (options[i].includes(e.target.value)) {
          if (e.target.value == correctAnswer[i]) {
            newScore[i]=1
            setScore(newScore)
          }
          else {
            newScore[i]=0
            setScore(newScore)
          }
        }
      }
    }

    const handleStart = async (data) => {
      const response = await axios.get(baseUrl+category)
      data = response.data.data
      const answers = [data.incorrect_answers_1.split(',')]
      answers[0].push(data.correct_answer_1)
      answers.push(data.incorrect_answers_2.split(','))
      answers[1].push(data.correct_answer_2)
      answers.push(data.incorrect_answers_3.split(','))
      answers[2].push(data.correct_answer_3)
      answers.push(data.incorrect_answers_4.split(','))
      answers[3].push(data.correct_answer_4)
      answers.push(data.incorrect_answers_5.split(','))
      answers[4].push(data.correct_answer_5)
      answers.push(data.incorrect_answers_6.split(','))
      answers[5].push(data.correct_answer_6)      
      console.log(answers)
      setOptions(answers)
      setQuestions([data.question_1, data.question_2, data.question_3, data.question_4, data.question_5, data.question_6])
      setCorrectAnswer([data.correct_answer_1, data.correct_answer_2, data.correct_answer_3, data.correct_answer_4, data.correct_answer_5, data.correct_answer_6])
      
      return data
    }
    const handleSubmit = (e) => {
      e.preventDefault()
      setFinalScore(score.reduce((a,b) => a+b))
    } 
   

  return (
    <div style={{ padding: "40px", paddingBottom: "100px" }}>
    <Paper>
    <FormControl>
      <FormControl>

        <InputLabel>Category</InputLabel>
        <Select 
          onChange={handleCategory}
          value={category}
          label="Category">
          <MenuItem value='CPR'>CPR</MenuItem>
          <MenuItem value='AED'>AED</MenuItem>
          <MenuItem value='Choking'>Choking</MenuItem>
        </Select>
      </FormControl>
      <FormControl>
        <Button variant="contained" onClick={handleStart} >Start Quiz</Button>

      </FormControl>
    </FormControl>
    </Paper>
   
    
         <div>
         
          <div>
            <form>    

              {
                !questions[0] ? '': questions.map((question, i) => {
                  return (
                  <div>
                    <h3>
                      Question {i+1}: {question}
                    </h3>
                    <input id={`q${i}o1`} type='radio' name={`question_${i}`} value={options[i][0]} onChange={handleScore}/>
                    <label for={`q${i}o1`}>{options[i][0]}</label>
                    <input id={`q${i}o2`} type='radio' name={`question_${i}`} value={options[i][1]} onChange={handleScore}/>
                    <label for={`q${i}o2`}>{options[i][1]}</label>
                    <input id={`q${i}o3`} type='radio' name={`question_${i}`} value={options[i][2]} onChange={handleScore}/>
                    <label for={`q${i}o3`}>{options[i][2]}</label>
                    <input id={`q${i}o4`} type='radio' name={`question_${i}`} value={options[i][3]} onChange={handleScore}/>
                    <label for={`q${i}o4`}>{options[i][3]}</label>
                    <br/>
                    <br/>
                    {i == 5 ? <Button variant='contained' type='submit' onClick={handleSubmit}>Submit</Button> :''}
                  </div>
                   )
                  })
              }
            </form>
          </div>
          <div>
            {finalScore ? <h3>Results: {finalScore}/{questions.length}</h3> : ""}
          </div>
         </div>
    </div>
  )
}


export default Quiz