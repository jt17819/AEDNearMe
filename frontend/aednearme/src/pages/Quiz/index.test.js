import { render, fireEvent, screen, cleanup, within } from '@testing-library/react';
import { MemoryRouter, Router as Router } from 'react-router-dom';
import ReactDOM from 'react-dom';
import { createMemoryHistory } from 'history'
import { default as Quiz } from './Quiz';
import axios from 'axios'

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe('Quiz', () => {
    let handleStartMock;

  // beforeEach(() => {
  //   render(<App />, { wrapper: MemoryRouter })
  // });


  // Rendering

  it('renders CPR button', () => {
    render(<Quiz />, { wrapper: MemoryRouter })
    expect(screen.getByRole("button", { name: "CPR" })).toBeInTheDocument();
  });

  it('renders Start Quiz button', () => {
    render(<Quiz />, { wrapper: MemoryRouter })
    expect(screen.getByRole("button", { name: "Start Quiz" })).toBeInTheDocument();
  });

//   it('renders results heading', () => {
//     render(<Quiz />, { wrapper: MemoryRouter })
//     expect(screen.getByRole("heading", { name: "Results: 0/0" })).toBeInTheDocument();
//   });

  test('list contains 2 buttons', () => {
    render(<Quiz />, { wrapper: MemoryRouter });
  
    // const listElement = screen.getAllByRole('button');
    const listItems = screen.getAllByRole('button');  
    expect(listItems.length).toEqual(2);
  });


    test('it calls on handleStart prop on form click', () => {
        handleStartMock = jest.fn();
        render(<MemoryRouter><Quiz /></MemoryRouter>);
        let startQuizButton = screen.getByText('Start Quiz');
        userEvent.click(startQuizButton)
        expect(handleStartMock).toHaveBeenCalled();
    })



  // Snapshot

  it('should take a snapshot', () => {
    render(<Quiz />, { wrapper: MemoryRouter })
    const { asFragment } = render(<Quiz />, { wrapper: MemoryRouter })
    
    expect(asFragment(<Quiz />)).toMatchSnapshot()
  })

  

  
})

