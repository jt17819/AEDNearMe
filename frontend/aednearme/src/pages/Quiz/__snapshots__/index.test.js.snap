// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should take a snapshot 1`] = `
<DocumentFragment>
  <div
    style="padding: 40px 40px 100px 40px;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink="true"
          >
            Category
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              CPR
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="CPR"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiFormControl-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Start Quiz
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div>
        <form />
      </div>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
